src/docker/tools/index.go

```diff
@@ -134,38 +136,41 @@ func TagImage(srcImage, targetImage string) error {
 }
 
 // PushImage pushes image with imageName to the registry given an encoded auth object
-func PushImage(c *client.Client, imageName, encodedAuth string) error {
+func PushImage(c *client.Client, writer io.Writer, imageName, encodedAuth string) error {
 	ctx := context.Background()
 	stream, err := c.ImagePush(ctx, imageName, dockerTypes.ImagePushOptions{
 		RegistryAuth: encodedAuth,
 	})
 	if err != nil {
 		return err
 	}
-	bytes, err := ioutil.ReadAll(stream)
-	if err != nil {
-		return err
+	scanner := bufio.NewScanner(stream)
+	for scanner.Scan() {
+		err = printPushProgress(writer, scanner.Text())
+		if err != nil {
+			return fmt.Errorf("Cannot push image '%s': %s", imageName, err)
+		}
 	}
-	err = parseDockerError(bytes)
-	if err != nil {
-		return err
+	if err := scanner.Err(); err != nil {
+		return errors.Wrap(err, "error reading ImagePush output")
 	}
 	return stream.Close()
 }
 
-func parseDockerError(output []byte) error {
-	outputArr := strings.Split(string(output), "\n")
-	errorMessage := outputArr[len(outputArr)-2]
-	errorObject := struct {
+func printPushProgress(writer io.Writer, output string) error {
+	outputObject := struct {
+		Status      string      `json:"status"`
+		ID          string      `json:"id"`
 		ErrorDetail interface{} `json:"errorDetail"`
 		Error       string      `json:"error"`
 	}{}
```

Comment:
Can we define this type in its own file? maybe named something like `PushImageProgress`?

Reply:
Do you think that's useful since it's not used anywhere else besides in this function?

Reply:
Good as is.  This is the first time I've seen this.